quvi.1
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "quvi 1"
.TH quvi 1 "2010-09-01" "0.2.3" "quvi manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
quvi \- query video tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
quvi [options] [url...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
quvi is a command line tool for parsing video download links.
It supports Youtube and other similar video websites.
.PP
The video details are printed to standard output, everything
else is printed to standard error.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-h, \-\-help\fR" 4
.IX Item "-h, --help"
Print help and exit.
.IP "\fB\-v, \-\-version\fR" 4
.IX Item "-v, --version"
Print program version and exit.
.IP "\fB\-\-hosts\fR" 4
.IX Item "--hosts"
Print supported hosts and exit.
.IP "\fB\-\-xml\fR" 4
.IX Item "--xml"
Print details in \s-1XML\s0. The details are escaped.
.IP "\fB\-q, \-\-quiet\fR" 4
.IX Item "-q, --quiet"
Turn off all output.
.IP "\fB\-\-debug\fR" 4
.IX Item "--debug"
Turn on libcurl verbose mode.
.IP "\fB\-n, \-\-no\-verify\fR" 4
.IX Item "-n, --no-verify"
Do not verify video link. Note that quvi will not report
video file length, content-type or file suffix if this
option is used.
.IP "\fB\-\-page\-title\fR=\fIstring\fR" 4
.IX Item "--page-title=string"
Match parsed video page title to \fIstring\fR.
.IP "\fB\-\-video\-id\fR=\fIstring\fR" 4
.IX Item "--video-id=string"
Match parsed video \s-1ID\s0 to \fIstring\fR.
.IP "\fB\-\-file\-length\fR=\fIlength\fR" 4
.IX Item "--file-length=length"
Match parsed video file length to \fIlength\fR.
.IP "\fB\-\-file\-suffix\fR=\fIstring\fR" 4
.IX Item "--file-suffix=string"
Match parsed video file suffix to \fIstring\fR.
.IP "\fB\-a, \-\-test\-all\fR" 4
.IX Item "-a, --test-all"
Run all built-in tests. quvi will test all
built-in video page links.
.IP "\fB\-t, \-\-test\fR=\fIpattern\fR" 4
.IX Item "-t, --test=pattern"
Match \fIpattern\fR to a built-in test page link
and test it.
.IP "\fB\-d, \-\-dump\fR" 4
.IX Item "-d, --dump"
Dump video details when running tests. To be used
with \fB\-\-test\-all\fR.
.IP "\fB\-\-agent\fR=\fIstring\fR" 4
.IX Item "--agent=string"
Identify quvi as \fIstring\fR to servers. Default
is \*(L"Mozilla/5.0\*(R".
.IP "\fB\-\-proxy\fR=\fIhost[:port]\fR" 4
.IX Item "--proxy=host[:port]"
Use the specified \s-1HTTP\s0 proxy. Note that quvi will
use http_proxy environment setting if it is set.
.IP "\fB\-\-connect\-timeout\fR=\fIseconds\fR" 4
.IX Item "--connect-timeout=seconds"
Maximum seconds allowed connection to server take.
Default is 30.
.IP "\fB\-f, \-\-format\fR[=]\fIformat_id\fR" 4
.IX Item "-f, --format[=]format_id"
Query video format, \fIformat_id\fR depends on the host,
see \fB\-\-hosts\fR output.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.IP "quvi \s-1URL\s0" 4
.IX Item "quvi URL"
Test \s-1URL\s0.
.IP "quvi \-a" 4
.IX Item "quvi -a"
Run all built-in host tests.
.IP "quvi \-aqd \-\-xml" 4
.IX Item "quvi -aqd --xml"
Same but run in quiet mode. Print video details in \s-1XML\s0 (to stdout).
.IP "quvi \-t youtube" 4
.IX Item "quvi -t youtube"
Match 'youtube' string to built-in video page links
and test it.
.IP "quvi \s-1URL\s0 \-f best" 4
.IX Item "quvi URL -f best"
Query 'best' of the video from the \s-1URL\s0.
.SH "BUGS"
.IX Header "BUGS"
Sure to be some. Please report them.
.PP
.Vb 1
\&  <http://quvi.googlecode.com/>
.Ve
.SH "DEBUGGING"
.IX Header "DEBUGGING"
.IP "quvi \-\-debug \s-1URL\s0" 4
.IX Item "quvi --debug URL"
Spew out libcurl verbose mode messages.
.ie n .IP "$prefix/share/quvi/lua/README" 4
.el .IP "\f(CW$prefix\fR/share/quvi/lua/README" 4
.IX Item "$prefix/share/quvi/lua/README"
.PD 0
.ie n .IP "$prefix/share/quvi/lua/website/README" 4
.el .IP "\f(CW$prefix\fR/share/quvi/lua/website/README" 4
.IX Item "$prefix/share/quvi/lua/website/README"
.PD
Contain details about working with the quvi Lua scripts (e.g. website
rules). If video link parsing breaks, have a look at these files.
.ie n .IP "$prefix/share/doc/quvi" 4
.el .IP "\f(CW$prefix\fR/share/doc/quvi" 4
.IX Item "$prefix/share/doc/quvi"
Various HOWTOs can be found in this directory.
.SH "FILES"
.IX Header "FILES"
.ie n .IP "\fB\fB$HOME\fB/.quvirc\fR" 4
.el .IP "\fB\f(CB$HOME\fB/.quvirc\fR" 4
.IX Item "$HOME/.quvirc"
You can define most of the command line options in the
config file. For example:
.Sp
.Vb 4
\& agent = some_agent/1.0     # \-\-agent
\& proxy = http://foo:1234    # \-\-proxy
\& no\-verify                  # \-\-no\-verify
\& debug                      # \-\-debug
.Ve
.Sp
Note that you can also define \s-1QUVI_HOME\s0 and use it instead.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\fB\s-1QUVI_HOME\s0\fR" 4
.IX Item "QUVI_HOME"
Defines path to a directory containing the configuration file.
Similar to \f(CW$HOME\fR found on Unix-like systems, but can be used
to override the \f(CW$HOME\fR setting (if any).
.IP "\fB\s-1QUVI_BASEDIR\s0\fR" 4
.IX Item "QUVI_BASEDIR"
Defines an exlusive path to the directory from which quvi searches
for the essential quvi files, e.g. the lua scripts. Exclusive here
means that by setting \fB\s-1QUVI_BASEDIR\s0\fR the user can override all of
the default search paths.
.Sp
Not to be confused with \fB\s-1QUVI_HOME\s0\fR which is strictly a \f(CWquvi(1)\fR
variable whereas \fB\s-1QUVI_BASEDIR\s0\fR is a libquvi variable.
.IP "\fB\s-1QUVI_SHOW_SCANDIR\s0\fR" 4
.IX Item "QUVI_SHOW_SCANDIR"
You can have libquvi dump (to stderr) the scanned directory
paths by setting this environment variable.
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
quvi exits with 0 on success and >0 if an error occurred.
.PP
.Vb 10
\&  QUVI_OK               = 0x00
\&  QUVI_MEM              = 0x01, Memory allocation failed
\&  QUVI_BADHANDLE        = 0x02, Bad session handle
\&  QUVI_INVARG           = 0x03, Invalid function argument
\&  QUVI_CURLINIT         = 0x04, libcurl initialization failed
\&  QUVI_LAST             = 0x05, Indicates end of list iteration
\&  QUVI_ABORTEDBYCALLBACK= 0x06, Aborted by callback function
\&  QUVI_LUAINIT          = 0x07, Lua initialization failure
\&  QUVI_NOLUAWEBSITE     = 0x08, Failed to find lua website scripts
\&  \-\-
\&  QUVI_PCRE             = 0x40, libpcre error occurred
\&  QUVI_NOSUPPORT        = 0x41, libquvi does not support the video host
\&  QUVI_CURL             = 0x42, libcurl error occurred
\&  QUVI_ICONV            = 0x43, libiconv error occurred
\&  QUVI_LUA              = 0x44, lua error occurred
.Ve
.SH "OTHER"
.IX Header "OTHER"
.IP "Project page:" 4
.IX Item "Project page:"
<http://quvi.googlecode.com/>
.IP "\s-1FAQ:\s0" 4
.IX Item "FAQ:"
<http://code.google.com/p/quvi/wiki/FAQ>
.IP "Development code:" 4
.IX Item "Development code:"
% git clone git://repo.or.cz/quvi.git
.SH "AUTHOR"
.IX Header "AUTHOR"
Toni Gundogdu.


NEWS

Version 0.2.0
=============

Contents

  1. Notes
  2. Introducing Lua
  3. Format ID changes
  4. API changes
  5. Distribution changes


1. Notes
--------

Developers:
* Read these changes carefully, especially "4. API changes"

Packagers/downstream maintainers/whatnot:
* Read (at least) the "5. Distribution changes"


2. Introducing Lua
------------------

Introduce Lua <http://lua.org> scripts to the project. Video link parsing
was completely reworked. All of the parsing is now done in website specific
Lua scripts. This has some advantages to the old design:

  * No need to tweak the library GNU Autotools files or C source code
  * Scripts can be edited, dropped in, removed and reloaded
  * Less likely to introduce memory leaks
  * Lua language and syntax


3. Format ID changes
--------------------

Common:
* "flv" -> "default", not all websites default to "flv" (Adobe flash video)

Youtube:
* Reworked the format ID array

Google (Video):
* Removed "mp4" format ID

Spiegel:
* Reworked the format ID array
* New default (vp6_64 -> vp_388)


4. API changes
--------------

Changes:
* Added additional QUVIcodes for Lua
* Added quvi_free
* Added quvi_next_supported_website
  - Deprecates quvi_next_host (see below)

Deprecated:
* quvi_next_host


5. Distribution changes
-----------------------

Changes:
* Additional dependency for Lua 5.1 (library)
* Additional files installed from doc/ and share/
  - See doc/Makefile.am and share/Makefile.am for a complete list
* Options for configure:
  - --with-doc (default:yes)
  - --enable-examples (default:no)


Version 0.1.2
=============

Adds quvi_callback_write support, tweaks quvi_callback_status
support (abort if the callback returns a non-zero value).

Adds --enable-broken.

Tests now include use of "-f best" where applicable.

Minor (unlisted) tweaks and bugfixes to quvi tool.

Marks as broken: ehrensenf.de, spyfilms.com, see TODO.


Version 0.1.1
-------------

Fixes the issues that have been plaguing the users on 64bit
systems. See ChangeLog for the details.

Dailymotion format IDs were renamed to reflect the website
changes. See ChangeLog for the details.

Fixes also Vimeo support.


src/quvi.c
/* 
* Copyright (C) 2009,2010 Toni Gundogdu.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
* quvi.c - query video tool.
*/

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <assert.h>

#include <curl/curl.h>
#include <pcre.h>

#include "platform.h"

#include "quvi/quvi.h"
#include "cmdline.h"

static int verbose_flag = 1;

static void
spew (const char *fmt, ...) {
    va_list ap;

    if (!verbose_flag)
        return;

    va_start (ap, fmt);
    vfprintf (stderr, fmt, ap);
    va_end (ap);
}

static void
spew_e (const char *fmt, ...) {
    va_list ap;
    va_start (ap, fmt);
    vfprintf (stdout, fmt, ap);
    va_end (ap);
}

typedef struct gengetopt_args_info opts_s;

static int
status_callback(long param, void *data) {
    quvi_word status,type;

    status = quvi_loword(param);
    type  = quvi_hiword(param);

    switch (status) {
    case QUVISTATUS_FETCH: /* handle fetch */
        switch (type) {
        default: spew(":: Fetch %s ...", (char *)data); break;
        case QUVISTATUSTYPE_CONFIG  : spew(":: Fetch config ..."); break;
        case QUVISTATUSTYPE_PLAYLIST: spew(":: Fetch playlist ..."); break;
        case QUVISTATUSTYPE_DONE    : spew("done.\n"); break;
        }
        break;

    case QUVISTATUS_VERIFY: /* handle verify */
        switch (type) {
        default: spew(":: Verify video link ..."); break;
        case QUVISTATUSTYPE_DONE: spew("done.\n"); break;
        }
        break;
    }

    fflush (stderr);

    return (0);
}

static size_t
write_callback (void *p, size_t size, size_t nmemb, void *data) {
    size_t r = quvi_write_callback_default(p, size, nmemb, data);
    /* Could do something useful here. */
#ifdef _1_
    puts(__func__);
#endif
    return r;
}

static const char notice[] =
"Copyright (C) 2009,2010 Toni Gundogdu. License: GNU GPL version  3  or later\n"
"This is free software; see the  source for  copying conditions.  There is NO\n"
"warranty;  not even for MERCHANTABILITY or FITNESS FOR A  PARTICULAR PURPOSE.";

static void
license (opts_s opts) {
    printf("%s\n",notice);
    cmdline_parser_free(&opts);
    exit (0);
}

static void
version (opts_s opts) {
    printf("quvi, version %s\n", quvi_version(QUVI_VERSION_LONG));
    cmdline_parser_free(&opts);
    exit (0);
}

static void
supports(quvi_t quvi, opts_s opts) {
    int done = 0;

    while (!done) {
        char *domain, *formats;
        QUVIcode rc = quvi_next_supported_website (quvi, &domain, &formats);

        switch (rc) {
        case QUVI_OK:
            printf("%s\t%s\n", domain, formats);
            quvi_free(domain);
            quvi_free(formats);
            break;
        case QUVI_LAST: done = 1; break;
        default       : fprintf(stderr, "%s\n", quvi_strerror(quvi, rc)); break;
        }
    }

    quvi_close(&quvi);
    cmdline_parser_free(&opts);

    exit (0);
}

static void
dump_video_links (quvi_video_t video, opts_s opts, CURL *curl) {
    const int xml = opts.xml_given;
    int i = 0;
    do {
        char *video_url, *file_suffix, *file_ct;
        double file_length;

        quvi_getprop(video, QUVIPROP_VIDEOURL, &video_url);
        quvi_getprop(video, QUVIPROP_VIDEOFILECONTENTTYPE, &file_ct);
        quvi_getprop(video, QUVIPROP_VIDEOFILESUFFIX, &file_suffix);
        quvi_getprop(video, QUVIPROP_VIDEOFILELENGTH, &file_length);

        if (!xml) {
            spew_e (
                "link %02d  : %s\n"
                ":: length: %.0f\n:: suffix: %s\n:: content-type: %s\n\n",
                ++i, video_url, file_length, file_suffix, file_ct
            );
        }
        else {
            char *ct  = curl_easy_escape (curl, file_ct, 0);
            char *url = curl_easy_escape (curl, video_url, 0);
            spew_e (
                "   <link id=\"%d\">\n"
                "       <length_bytes>%.0f</length_bytes>\n"
                "       <content_type>%s</content_type>\n"
                "       <file_suffix>%s</file_suffix>\n"
                "       <url>%s</url>\n"
                "   </link>\n",
                ++i, file_length, ct, file_suffix, url
            );
            curl_free (url);
            url = NULL;
            curl_free (ct);
            ct = NULL;
        }
    } while (quvi_next_videolink(video) == QUVI_OK);
}

static void
dump_video(quvi_video_t video, opts_s opts, CURL *curl) {
    char *page_link, *page_title, *video_id, *format;
#ifdef _1_
    long httpcode;
#endif

    quvi_getprop(video, QUVIPROP_PAGEURL, &page_link);
    quvi_getprop(video, QUVIPROP_PAGETITLE, &page_title);
    quvi_getprop(video, QUVIPROP_VIDEOID, &video_id);
    quvi_getprop(video, QUVIPROP_VIDEOFORMAT, &format);
#ifdef _1_
    quvi_getprop(video, QUVIPROP_HTTPCODE, &httpcode);
#endif

    if (!opts.xml_given) {
        spew_e (
            " > Dump video:\n"
            "url     : %s\n"
            "title   : %s\n"
            "id      : %s\n"
            "format  : %s (requested)\n",
            page_link, page_title, video_id, format
        );
    }
    else {
        char *id    = curl_easy_escape (curl, video_id, 0);
        char *title = curl_easy_escape (curl, page_title, 0);
        char *url   = curl_easy_escape (curl, page_link, 0);
        spew_e (
        "<?xml version='1.0' encoding='utf-8'?>\n"
        "<video id=\"%s\">\n"
        "   <format_requested>%s</format_requested>\n"
        "   <page_title>%s</page_title>\n"
        "   <page_url>%s</page_url>\n",
        id, format, title, url
        );
        curl_free (url);
        url = NULL;
        curl_free (title);
        title = NULL;
        curl_free (id);
        id = NULL;
    }

    dump_video_links (video, opts, curl);

    if (opts.xml_given)
        spew_e ("</video>\n");

#ifdef _0
    spew("httpcode: %ld (last)\n", httpcode);
#endif
}

static void
expect_error (const char *what, const char *expected, const char *got) {
    fprintf(stderr,
        "error: %s:\n  expected: \"%s\"\n  got: \"%s\"\n\n",
            what, expected, got);
}

static void
expect_error_d (const char *what, const double expected, const double got) {
    fprintf(stderr,
        "error: %s:\n  expected: \"%.0f\"\n  got: \"%.0f\"\n\n",
            what, expected, got);
}

static int
check_values(quvi_video_t video, opts_s opts) {
    int rc = 0;

    if (opts.page_title_given) {
        char *title, *expect=opts.page_title_arg;
        quvi_getprop(video, QUVIPROP_PAGETITLE, &title);
        rc = strcmp(expect, title) != 0;
        if (rc)
            expect_error("page title", expect, title);
    }

    if (opts.video_id_given && !rc) {
        char *id, *expect=opts.video_id_arg;
        quvi_getprop(video, QUVIPROP_VIDEOID, &id);
        rc = strcmp(expect, id) != 0;
        if (rc)
            expect_error("video id", expect, id);
    }

    if (opts.file_suffix_given && !rc) {
        char *suffix, *expect=opts.file_suffix_arg;
        quvi_getprop(video, QUVIPROP_VIDEOFILESUFFIX, &suffix);
        rc = strcmp(expect, suffix) != 0;
        if (rc)
            expect_error("file suffix", expect, suffix);
    }

    if (opts.file_length_given && !rc) {
        double length, expect=opts.file_length_arg;
        quvi_getprop(video, QUVIPROP_VIDEOFILELENGTH, &length);
        rc = expect != length;
        if (rc)
            expect_error_d("file length", expect, length);
    }

    return (rc);
}

static void
dump_error(quvi_t quvi, QUVIcode rc, opts_s opts) {
    long httpcode, curlcode;

    quvi_getinfo(quvi, QUVIINFO_HTTPCODE, &httpcode);
    quvi_getinfo(quvi, QUVIINFO_CURLCODE, &curlcode);

    fprintf(stderr, "\nerror: %s (http/%ld, curl/%ld)\n",
        quvi_strerror(quvi,rc),
        httpcode, curlcode);

    if (!opts.test_all_given) {
        quvi_close(&quvi);
        cmdline_parser_free(&opts);
        exit (rc);
    }
}

static const char *tests[] = {
"http://www.dailymotion.com/hd/video/x9fkzj_battlefield-1943-coral-sea-trailer_videogames",
"http://www.spiegel.de/video/video-1012582.html",
"http://vimeo.com/1485507",
"http://en.sevenload.com/videos/IUL3gda-Funny-Football-Clips",
"http://www.myubo.com/page/media_detail.html?movieid=1308f0fb-47c6-40c5-a6f9-1324dac12896",
"http://www.liveleak.com/view?i=704_1228511265",
"http://video.google.com/videoplay?docid=-8669127848070159803",
"http://video.golem.de/internet/2174/firefox-3.5-test.html",
"http://www.funnyhub.com/videos/pages/crazy-hole-in-one.html",
"http://www.clipfish.de/video/3100131/matratzendomino/",
"http://www.youtube.com/watch?v=DeWsZ2b_pK4",
"http://break.com/index/beach-tackle-whip-lash.html",
"http://www.evisor.tv/tv/rennstrecken/1-runde-oschersleben-14082008--6985.htm",
"http://www.buzzhumor.com/videos/32561/Girl_Feels_Shotgun_Power",
"http://www.funnyordie.com/videos/776d200b1c/etiquette-ninjas-episode-5-dicks-on-elevators",
"http://www.charlierose.com/view/interview/11125",
"http://www.theonion.com/video/time-announces-new-version-of-magazine-aimed-at-ad,17950/",

#ifdef ENABLE_BROKEN
"http://space.tv.cctv.com/video/VIDE1212909276513233", /* single-segment */
"http://space.tv.cctv.com/video/VIDE1247468077860061", /* multi-segment */
#endif

#ifdef ENABLE_SMUT
"http://www.tube8.com/fetish/japanese-melon-gal-censored/186133/",
"http://www.xvideos.com/video243887/devi_emmerson_body_painting",
"http://www.youjizz.com/videos/glamour-girls---melissa-125602.html",
#endif

NULL
};

static void
match_test (quvi_t quvi, opts_s opts, CURL *curl) {
    unsigned register int i;
    const char *errmsg;
    int err_offset, rc;

    for (i=0; tests[i]; ++i) {
        pcre *re;

        re = pcre_compile(
            opts.test_arg,
            PCRE_CASELESS,
            &errmsg,
            &err_offset,
            NULL
        );

        if (!re) {
            fprintf(stderr, "%s\n", errmsg);
            cmdline_parser_free(&opts);
            exit (QUVI_PCRE);
        }

        rc = pcre_exec(
            re,
            0,
            tests[i],
            strlen(tests[i]),
            0,
            0,
            0,
            0
        );

        pcre_free(re);

        if (rc >= 0) /* match */ {
            quvi_video_t v;
            QUVIcode _rc;

            _rc = quvi_parse(quvi, (char *)tests[i], &v);
            if (_rc != QUVI_OK)
                dump_error(quvi, _rc, opts);

            dump_video (v, opts, curl);
            rc = check_values(v, opts);
            quvi_parse_close(&v);

            cmdline_parser_free(&opts);
            quvi_close(&quvi);
            exit (rc);
        }
        else if (rc == PCRE_ERROR_NOMATCH) {
            continue;
        }
        else {
            fprintf(stderr, "error: pcre_exec: rc = %d\n", rc);
            cmdline_parser_free(&opts);
            exit (QUVI_PCRE);
        }
    }

    fprintf(stderr, "error: nothing matched `%s'\n", opts.test_arg);
    cmdline_parser_free(&opts);
    quvi_close(&quvi);
    exit (QUVI_PCRE);
}

static void
test_all (quvi_t quvi, opts_s opts, CURL *curl) {
    register unsigned int i,m;
    quvi_video_t video;
    QUVIcode rc;

    spew(":: Run built-in video link tests.\n");
    for (m=0; tests[m]; ++m);

    for (i=0; i<m; ++i) {
        spew(" > Test #%02d/%02d:\n", i+1, m);

        rc = quvi_parse(quvi, (char *)tests[i], &video);
        if (rc != QUVI_OK)
            dump_error(quvi, rc, opts);
        else {
            if (opts.dump_given)
                dump_video (video, opts, curl);
        }
        quvi_parse_close(&video);
    }
    spew(":: Tests done.\n");
}

static quvi_t
init_quvi(opts_s opts, CURL **curl) {
    QUVIcode rc;
    quvi_t quvi;

    if ( (rc = quvi_init(&quvi)) != QUVI_OK )
        dump_error(quvi, rc, opts);
    assert(quvi != 0);

    /* Set quvi options. */

    if (opts.format_given)
        quvi_setopt(quvi, QUVIOPT_FORMAT, opts.format_arg);

    if (opts.no_verify_given)
        quvi_setopt(quvi, QUVIOPT_NOVERIFY, 1L);

    quvi_setopt(quvi, QUVIOPT_STATUSFUNCTION, status_callback);
    quvi_setopt(quvi, QUVIOPT_WRITEFUNCTION, write_callback);

    /* We can use the quvi created cURL handle for our means. */

    quvi_getinfo(quvi, QUVIINFO_CURL, curl);
    assert(*curl != 0);

    if (opts.agent_given)
        curl_easy_setopt (*curl, CURLOPT_USERAGENT, opts.agent_arg);

    if (opts.proxy_given)
        curl_easy_setopt (*curl, CURLOPT_PROXY, opts.proxy_arg);

    if (opts.debug_given)
        curl_easy_setopt (*curl, CURLOPT_VERBOSE, 1L);

    curl_easy_setopt (*curl,
        CURLOPT_CONNECTTIMEOUT, opts.connect_timeout_arg);

    return (quvi);
}

int
main (int argc, char *argv[]) {
    const char *url, *home, *no_config, *fname;
    register unsigned int i;
    quvi_video_t video;
    int no_config_flag;
    opts_s opts;
    QUVIcode rc;
    quvi_t quvi;
    CURL *curl;

    curl      = NULL;
    url       = NULL;
    no_config = getenv("QUVI_NO_CONFIG");
    home      = getenv("QUVI_HOME");
    if (!home)
        home = getenv("HOME");
    no_config_flag   = 1;

#ifndef HOST_W32
    fname = "/.quvirc";
#else
    fname = "\\quvirc";
#endif

    /* Init cmdline parser. */

    if (home && !no_config) {
        char *path;
        FILE *f;

        asprintf(&path, "%s%s", home, fname);
        f = fopen(path, "r");

        if (f != NULL) {
            struct cmdline_parser_params *pp;

            fclose(f);
            f = NULL;

            pp = cmdline_parser_params_create();
            pp->check_required = 0;

            if (cmdline_parser_config_file(path, &opts, pp) == 0) {
                pp->initialize     = 0;
                pp->override       = 1;
                pp->check_required = 1;

                if (cmdline_parser_ext(argc, argv, &opts, pp) == 0)
                    no_config_flag = 0;
            }
            free(pp);
            pp = NULL;
        }

        free(path);
        path = NULL;
    }

    if (no_config_flag) {
        if (cmdline_parser(argc, argv, &opts) != 0)
            return (QUVI_INVARG);
    }

    /* Run and forget options. */

    if (opts.version_given)
        version(opts);

    if (opts.license_given)
        license(opts);

    /* Other */

    verbose_flag = !opts.quiet_given;

    /* Init quvi. */

    quvi = init_quvi (opts, &curl);

    /* --hosts */
    if (opts.hosts_given)
        supports(quvi, opts);

    /* Handle input. */
    if (opts.test_all_given)
        test_all (quvi, opts, curl);
    else if (opts.test_given)
        match_test(quvi, opts, curl);
    else
    {
        if (opts.inputs_num == 0)
            fprintf(stderr, "error: no input links\n");

        for (i=0; i<opts.inputs_num; ++i)
        {
            if ( (rc = quvi_parse(
                    quvi,
                    (char *)opts.inputs[i],
                    &video))
                != QUVI_OK )
            {
                dump_error(quvi, rc, opts);
            }

            assert(video != 0);
            dump_video(video, opts, curl);

            quvi_parse_close(&video);
            assert(video == 0);
        }
    }

    /* Cleanup. */

    quvi_close(&quvi);
    assert(quvi == 0);

    cmdline_parser_free(&opts);

    return (QUVI_OK);
}


